ocaml/utils/misc.ml

```diff
@@ -993,24 +994,34 @@ module EnvLazy = struct
     match !x with
     | Thunk a -> Either.Left a
     | Done b -> Either.Right b
-    | Raise e -> raise e
+    | Raise (e, bt) -> Printexc.raise_with_backtrace e bt
 
   let create x =
     ref (Thunk x)
 
   let create_forced y =
     ref (Done y)
 
+  type does_not_return = |
+  exception I_just_want_a_backtrace
+
   let create_failed e =
-    ref (Raise e)
+    let bt =
+      (* We want to call [Printexc.get_raw_backtrace], but that's only valid in
+         an exception handler. So let's handle an exception. *)
+      match (raise I_just_want_a_backtrace : does_not_return) with
+      | exception _ -> Printexc.get_raw_backtrace ()
+      | _ -> .
+    in
```

Comment:
`Printexc.get_callstack` is what you want here

Reply:
Ah, nice - yes, clearly that's better (it's more sensible _and_ the output is better).